### ISSUE TITLE
Use Sparse Matrices

### ISSUE THREAD

**Issue by ahwillia** [madeleineudell/LowRankModels.jl#33]
I have a few applications that involve *very* large, *very* sparse (incomplete), matrices. 

At the moment it appears that the entire data matrix, `A`, is estimated/reconstructed from `X'*Y` on each gradient step ([see here](https://github.com/madeleineudell/LowRankModels.jl/blob/master/src/algorithms/proxgrad.jl#L37
)). However, if many entries of `A` are missing, then only a subset of the `X'*Y` entries need to be computed to evaluate the loss function.

Would it make sense to make a new file `sparse_proxgrad.jl` in `src/algorithms/` that is called when `glrm.A` is a sparse matrix?

**Comment by madeleineudell** [comment]
Yes, that would be great! I think you've hit on (basically) the only change
you'd need to make.

I think the right way to call it is to dispatch on the type of params;
params::SparseProxGradParams would dispatch to the sparse_proxgrad fitting
method. And then we could default to params=SparseProxGradParams() when
glrm.A is a sparse matrix.

On Tue, Aug 11, 2015 at 11:41 AM, Alex Williams <notifications@github.com>
wrote:

> I have a few applications that involve *very* large, *very* sparse
> (incomplete), matrices.
>
> At the moment it appears that the entire data matrix, A, is
> estimated/reconstructed from X'*Y on each gradient step (see here
> <https://github.com/madeleineudell/LowRankModels.jl/blob/master/src/algorithms/proxgrad.jl#L37>).
> However, if many entries of A are missing, then only a subset of the X'*Y
> entries need to be computed to evaluate the loss function.
>
> Would it make sense to make a new file sparse_proxgrad.jl in
> src/algorithms/ that is called when glrm.A is a sparse matrix?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/madeleineudell/LowRankModels.jl/issues/33>.
>



-- 
Madeleine Udell
Postdoctoral Fellow at the Center for the Mathematics of Information
California Institute of Technology
www.stanford.edu/~udell
(415) 729-4115

**Comment by ahwillia** [comment]
Ok -- I'll try to make a first-pass implementation of this over the next couple of days.

**Comment by ahwillia** [comment]
Looking into this now... I'm not sure if I'm missing something, but what would be wrong with replacing [this line](https://github.com/madeleineudell/LowRankModels.jl/blob/master/src/algorithms/proxgrad.jl#L72):

```julia
axpy!(grad(losses[f],XY[e,f],A[e,f]), vf[f], g)
```

as follows:

```julia
axpy!(grad(losses[f],dot(X[:,e],Y[:,f]),A[e,f]), vf[f], g)
```

or (probably better):

```julia
axpy!(grad(losses[f],dot(ve[e],vf[f]),A[e,f]), vf[f], g)
```

It seems like this would work just as well for both sparse and dense A, so we wouldn't need to define a new `fit!` function.

**Comment by madeleineudell** [comment]
If you're going to access (nearly) every entry in the product matrix, it's
faster to compute the whole product matrix up front. Not because it's fewer
flops (it isnt) but because of hardware level optimizations in LAPACK:
chucking the operations so they just fit in various caches, etc. So I'd
prefer to keep the first variant for dense and use the last variant for
sparse.

But it's worthwhile checking the argument I've made here empirically: for
large dense matrices, which is faster in practice?

**Comment by ahwillia** [comment]
Closed by #35
